### PR TITLE
Added create, next, entry commands.

### PR DESCRIPTION
**Pull request by vpassapera**
Added a few commands to make it easier to add and manipulate the changelog (without the need to use an editor).

Adds the following:

       create     Creates a new CHANGELOG.md file with a standard header.
        entry      Add a quick entry to the current version.
        next       Create a new unreleased version in changelog file.

Allows to change vendor on install script to be friendlier to downstream
repos and automation.

### CODE REVIEW

**Comment by vpassapera** [comment]
Went ahead and updated the README as well. Now this has updated readme, updated tests, and 3 new features that allow for fully automated management of CHANGELOG.

I'm currently using my downstream version along with jenkins and some git magic to automatically create a custom changelog based on git commits.

It's quite nice. Hoping this gets merged upstream, but if it doesn't I'll keep maintaining my downstream version.

**Comment by ralph-tice** [comment]
Just saw this, nice!  I ran into an error with
```
$ chag entry Added WAF example for intuit CIDR ingress rules
sed: -i may not be used with stdin
```

I'm on OSX, so I assume that's what's introducing the discrepancy.  Removing the `-i` from L255 gives a new error though:
```
$ chag entry Added WAF example for intuit CIDR ingress rules
sed: 1: "5i \* ": extra characters after \ at the end of i command
```

I'm a little strapped for time right now but I might circle back and poke at this some more, just wanted to capture the error for now.

**Comment by vpassapera** [comment]
The docs say to do chag entry -contents "<contents go here>"

**Comment by ralph-tice** [comment]
Oops!  Same problem though:
```
$ chag entry --contents "Added WAF example for intuit CIDR ingress rules"
sed: -i may not be used with stdin
```

I started writing tests but got stuck on some bashisms :)

**Comment by vpassapera** [comment]
Hmmm. Interesting. I do not currently have access to my osx machine, but will look into this.

**Comment by vpassapera** [comment]
@ralph-tice seems the problem is due to the bsd implementation of sed.

https://stackoverflow.com/questions/21242932/sed-i-may-not-be-used-with-stdin-on-mac-os-x

I might add a homebrew requirement for osx to install gnu-sed and then call that from within code instead of bsd sed to fix this problem. But once I go down this path I'll probably just maintain my own fork.

**Comment by ralph-tice** [comment]
that seems like a pretty heavy solution compared to https://stackoverflow.com/questions/21242932/sed-i-may-not-be-used-with-stdin-on-mac-os-x#comment36378028_21243111 no?

**Comment by ralph-tice** [comment]
I started writing some happy path tests here: https://github.com/ralph-tice/chag/blob/new_cmds/test/next.bats#L17-L22

Trying not to be too messy but I'd like to help 100% this effort you've started for initiating new changelogs :)

**Comment by vpassapera** [comment]
@ralph-tice if you want to PR into my repo, I always welcome that.

Thanks btw!

**Comment by dpsenner** [comment]
Superseded by #10.

**Comment by vpassapera** [comment]
I think that this repo is no longer maintained.... You could always open up the PR against my fork as well as here.

File: chag [hunk 1]
```diff
@@ -173,7 +213,7 @@ latest() {
 
 # Get the line number of a specific changelog entry.
 find_entry() {
-  result=$(entries_and_lines | grep -E ":${1//./\\.}$" | cut -d ':' -f 1)
+  result=$(entries_and_lines | grep "${1//./\\.}" | cut -d ':' -f 1)
```
Comment (mtdowling) [inline chag]: I asked this on the other PR. What is the motivation for this change?

Comment (vpassapera) [inline chag]: Mine didn't work after that line was changed to the new one. But I should try it again.

Comment (dpsenner) [inline chag]: For me both versions work just fine.

**Comment by vpassapera** [comment]
@mtdowling I went ahead and tested with your current grep command and it seems to work.

File: test/next.bats [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+load test_helper
+
+@test "--help prints help" {
+  run ./chag next --help
+  [ $status -eq 0 ]
+  [ $(expr "${lines[0]}" : "Usage: chag next") -ne 0 ]
+}
+
+@test "Invalid options fail" {
```
Comment (mtdowling) [inline test/next.bats]: Can you add a test here that tests for a working call to next?

File: test/create.bats [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bats
```
Comment (mtdowling) [inline test/create.bats]: Can you add a test here that tests for a working call to create?

File: README.rst [hunk 4]
```diff
@@ -116,6 +119,52 @@ Usage
...
+::
+
+   Usage: chag next
+
+   Creates a new unreleased version inside of the changelog file.
+
+   Options:
+     --help     Displays this message.
+
+entry
```
Comment (mtdowling) [inline README.rst]: I'm not sure the entry command is a good idea long term. The description is right that this is mostly just useful for quick entries to be added to the changelog; trying to add things like multi-line entries would not be easy.  I would much rather prefer people to be verbose and descriptive when editing their changelogs. I think most people will be editing their code when they are getting ready to add an entry to the changelog. They are typically in an editor of some kind, and I think that's the more appropriate place to edit the changelog. Looking at the way it's implemented, I'm also wary of taking on the complexity of supporting this for such a (IMO) limited use case.

Comment (vpassapera) [inline README.rst]: I use chag to automate building my changelog based on commit messages.

E.g. I iterate through my commits and add the commits in a given release to the changelog using chag. For this I use this command.

So people that want to do multiline probably wouldn't use the command if they don't want, but for automated purposes, it actually works.

Comment (dpsenner) [inline README.rst]: The `entry` command is synonymous to `entries` but does something entirely different. That's odd and should therefore probably read as `addentryitem` as that reflects more what this command does.

@mtdowling has a good point though that people rather use an editor to modify the changelog and that's also why I thought about adding a command `edit` that opens the systems default editor with the content of the latest entry and replaces the content of that entry with the edits made in that editor session. It would behave much alike `git commit` does.

Further the commands are not intuitive to me. The following is a summary of all suggestions:

* `create` could be `init` (alike `git init`)
* `entries` could be `list` or `showentries` or `listitems`
* `next` could be `add` or `addentry` or `additem` or `addnext` or `addunreleased`
* add new command `edit` or `editentry` or `edititem` or `editchanges`
* `contents` could be `show` or `showentry` or `showitem` or `showchanges`
* `update` could be `setversion` or `setentryversion` or `setitemversion`
* `entry` could be `addentryitem` or `addchange`

The first item is what I see as the best choice.

Comment (vpassapera) [inline README.rst]: @dpsenner you're suggesting quite a few changes that should probably be broken down across multiple PRs and would basically create a LOT of BC breaks.

Your suggestions are noted but i think are very much outside of the scope of this singple PR.

Thank you.

File: test/next.bats (hunk 2 above)
Comment (vpassapera) [inline test/next.bats]: I'll try to do something about this soon.

File: test/create.bats (hunk 3 above)
Comment (vpassapera) [inline test/create.bats]: Same here.

File: README.rst (hunk 4 above)
Comment (dpsenner) [inline README.rst]: So long as we alias the commands and deprecate the current ones there will be no breaking changes until release 2.0.0 is made. These suggestions are merely here to start a discussion and find a consensus before even think of a formal specification, yet not to talk about an implementation.

Comment (vpassapera) [inline README.rst]: I should also note that highjacking other people's PRs and then trying to "supersede" them with your own is generally poor open source etiquette. If you would like to build on my PR, one way of doing so is by rebasing your PR against mine if mine it is merged, or if you want to implement the same features but differently then you can do so, but not including my commits on your own and then commenting on my PR as if I agreed with the changes you added to my changes.

Thanks for your own source interest.

Comment (vpassapera) [inline README.rst]: As for the stuff you're suggesting for renaming commands, I appreciate what you're saying, but I do not think this is the appropriate PR for those changes.

I suggest you open up a new issue and suggest those changes to @mtdowling or do them in your fork and go from there.

Comment (dpsenner) [inline README.rst]: I thought your changes are worth to be pulled into my fork since I use them. I did not know you would not like that to happen and apologize for doing so. Rebasing and rewriting history regularly is not a trivial task and I like to save time where ever I can for stuff that has greater importance to me. So there's only one option that I can do at this point. Would you like me to back out all of your new commands from my fork?

Comment (vpassapera) [inline README.rst]: I am not suggesting that my changes should not be used by other people, what I am saying is that including my changes in your own PR and then claiming this one is in so many words "deprecated" is not polite because you never asked me if that was the case.

Considering this PR is still active and I'm actively making changes to it as they are requested by repo owner I think that it would have been ideal if you had either tried to collaborate to modify this PR or at least discussed your suggestions prior to just commenting on this with "Superseded by...".

I appreciate that you think these changes are good enough to want to use them (after all that's the spirit of open source!) but we should be polite in our approach to collaborating and not aggressive, so not to be misunderstood.

Besides the points I am making above, it seems that your PR is adding other unrelated functionality such as a config file, that really doesn't have much to do with what I am doing here but could be useful in it's own right.

Comment (vpassapera) [inline README.rst]: Btw, rebasing is fairly easy. git rebase <target>.

I am not sure about this repo's rules regarding cleanliness of commit messages, but most repos on GH have a rebase / squash policy. I suggest you familiarize yourself with those for future participation.

Comment (dpsenner) [inline README.rst]: Your pull request has not been merged after more than a year and that's why I make little to no assumptions whether this is going to happen soon or anytime at all. If I did not move on I would still wait for your changes to be merged into the original repository. Thus I would have to use two versions of chag to use all the features of both forks. This is one of the main reasons why forking and permissive licensing of creative work exist.

Because of that I pulled and merged your work into my fork because to me it looked good, it works and it eases my day to day work. I further improved `chag` and it would now be impossible to rebase again without producing merge conflicts. Given that all your work has been merged into the other fork and all the work in that fork is available as a pull request it is indeed the case that pull request #10 supersedes this pull request. There's no offense or aggressive behavior behind that. It simply states that the work of this pull request is included in another pull request. If you think that forking is "hijacking" (a synonym to "stealing") of your creative work, please re-read the MIT license. To me this discussion ends here and I am going back to work on something productive like improving the usability of chag.

Comment (dpsenner) [inline README.rst]: I created [this issue](https://github.com/dpsenner/chag/issues/2). I would like anyone invite to join the conversation.

Comment (vpassapera) [inline README.rst]: @dpsenner Vulturing PRs is generally only acceptable if the PR is not active, you can clearly see with the comments going on in here that, that is not the case and this PR is indeed active. The PR not being merged is because the maintainer did not respond to the PR for an extremely long time.

Once again, I did not say that people can't use my code, and once again, I am fully aware of what the MIT license allows. Your strawman argument falls appart when you look at the content of my comments THIS VERY YEAR, however, your extremely aggressive and obnoxious behavior is not excused by the MIT license.

The MIT license doesn't suggest you to try to get another PR closed so you can have yours merged after simply aggregating other's work and putting it together. You should familiarize yourself with collaboration etiquette as I suggested before.

Now, I would appreciate if you let me handle my PR without further interruptions or distractions unless you are explicitly contributing to things related to this PR specifically.

Thank you.
